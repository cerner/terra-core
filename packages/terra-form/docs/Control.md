# Control

A Control represents a radio or checkbox widget with a label. Uses the Input component.

## Usage

```jsx
import React from 'react';
import Control from 'terra-form/lib/Control';

<Control
  type="checkbox"
  defaultChecked={false}
  id="form-checkbox"
  inputAttrs={{ className: 'healtheintent-checkbox' }}
  labelText="Do you have any Children?"
  labelTextAttrs={{ className: 'healtheintent-control-label-text' }}
  name="children_present"
  onChange={() => {}}
  value="children_present"
  isInline
 />
```
