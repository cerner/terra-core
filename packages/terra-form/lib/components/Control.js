'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Input = require('./Input');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } // eslint-disable react/jsx-boolean-value

var propTypes = {
  /**
   * The input type
   */
  type: _react.PropTypes.oneOf(['checkbox', 'radio']).isRequired,
  /**
   * Additional attributes for the input object
   */
  attrs: _react.PropTypes.object,
  /**
   * id of the input field
   */
  id: _react.PropTypes.string,
  /**
   * Whether the form element is inline or not
   */
  isInline: _react.PropTypes.bool,
  /**
   * Label of the input
   */
  label: _react.PropTypes.node,
  /**
   * Name of the input attribute
   */
  name: _react.PropTypes.string,
  /**
   * Initial Value of the input
   */
  value: _react.PropTypes.string
};

var defaultProps = {
  type: null,
  attrs: {},
  id: undefined,
  isInline: false,
  label: null,
  name: null,
  value: undefined
};

var Control = function Control(_ref) {
  var type = _ref.type,
      attrs = _ref.attrs,
      id = _ref.id,
      isInline = _ref.isInline,
      label = _ref.label,
      name = _ref.name,
      value = _ref.value,
      customProps = _objectWithoutProperties(_ref, ['type', 'attrs', 'id', 'isInline', 'label', 'name', 'value']);

  var labelClassNames = (0, _classnames2.default)('terra-Form-control', { 'terra-Form-control--inline': isInline }, customProps.className);

  return _react2.default.createElement(
    'label',
    _extends({ htmlFor: id }, customProps, { className: labelClassNames }),
    _react2.default.createElement(_Input2.default, _extends({
      id: id,
      name: name,
      value: value,
      type: type
    }, attrs)),
    label
  );
};

Control.propTypes = propTypes;
Control.defaultProps = defaultProps;

exports.default = Control;