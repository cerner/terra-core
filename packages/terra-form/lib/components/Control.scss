@import '~terra-mixins';

.terra-Form-control {
  display: block;
  margin-bottom: 0.714em;
  width: auto;
}

@media screen and (min-width: $terra-small-breakpoint) {
  .terra-Form-control--inline {
    display: inline-block;
    vertical-align: top;
  }
}

.terra-Form-control {
  @include terra-margin-end(0.714em);
}
