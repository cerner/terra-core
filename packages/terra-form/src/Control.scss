@import '~terra-mixins';

.terra-Form-control {
  display: block;
  margin-bottom: 0.714em;
  width: auto;

  .terra-Form-input {
    vertical-align: middle;
  }
}

@media screen and (min-width: 544px) {
  .terra-Form-control--inline {
    display: inline-block;
    @include terra-margin-end(0.714em);
  }
}

.terra-Form-control-label-text {
  vertical-align: middle;
}
