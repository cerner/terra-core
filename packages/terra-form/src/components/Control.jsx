// eslint-disable react/jsx-boolean-value

import React, { PropTypes } from 'react';
import classNames from 'classnames';

import Input from './Input';

const propTypes = {
  /**
   * The input type
   */
  type: PropTypes.oneOf(['checkbox', 'radio']).isRequired,
  /**
   * Additional attributes for the input object
   */
  attrs: PropTypes.object,
  /**
   * id of the input field
   */
  id: PropTypes.string,
  /**
   * Whether the form element is inline or not
   */
  isInline: PropTypes.bool,
  /**
   * Label of the input
   */
  label: PropTypes.node,
  /**
   * Name of the input attribute
   */
  name: PropTypes.string,
  /**
   * Initial Value of the input
   */
  value: PropTypes.string,
};

const defaultProps = {
  type: null,
  attrs: {},
  id: undefined,
  isInline: false,
  label: null,
  name: null,
  value: undefined,
};

const Control = ({ type, attrs, id, isInline, label, name, value, ...customProps }) => {
  const labelClassNames = classNames(
    'terra-Form-control',
    { 'terra-Form-control--inline': isInline },
    customProps.className,
  );

  return (
    <label htmlFor={id} {...customProps} className={labelClassNames}>
      <Input
        id={id}
        name={name}
        value={value}
        type={type}
        {...attrs}
      />
      {label}
    </label>
  );
};

Control.propTypes = propTypes;
Control.defaultProps = defaultProps;

export default Control;
