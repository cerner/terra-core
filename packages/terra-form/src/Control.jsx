import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import 'terra-base/lib/baseStyles';

import Input from './Input';

import './Control.scss';

const propTypes = {
  /**
   * The input type
   */
  type: PropTypes.oneOf(['checkbox', 'radio']).isRequired,
  /**
   * The checked property of the Input element. Use this to generate a controlled Control Element
   */
  checked: PropTypes.bool,
  /**
   * The checked property of the Input element. Use this to generate an uncontrolled Control Element
   */
  defaultChecked: PropTypes.bool,
  /**
   * id of the input field
   */
  id: PropTypes.string,
  /**
   * Additional attributes for the input object
   */
  inputAttrs: PropTypes.object,
  /**
   * Whether the form element is inline or not
   */
  isInline: PropTypes.bool,
  /**
   * Text of the label
   */
  labelText: PropTypes.node,
  /**
   * Additional objects for the text object
   */
  labelTextAttrs: PropTypes.object,
  /**
   * Name of the input attribute
   */
  name: PropTypes.string,
  /**
   * Function to trigger when user clicks on the input. Provide a function to create a controlled input.
   */
  onChange: PropTypes.func,
  /**
   * The Value of the input element
   */
  value: PropTypes.string,
};

const defaultProps = {
  type: null,
  checked: undefined,
  defaultChecked: undefined,
  id: undefined,
  inputAttrs: {},
  isInline: false,
  labelText: null,
  labelTextAttrs: {},
  name: null,
  onChange: undefined,
  value: undefined,
};

const Control = ({
  type,
  checked,
  defaultChecked,
  inputAttrs,
  id,
  isInline,
  labelText,
  labelTextAttrs,
  name,
  onChange,
  value,
  ...customProps
}) => {
  const labelClassNames = classNames(
    'terra-Form-control',
    { 'terra-Form-control--inline': isInline },
    customProps.className,
  );

  const labelTextClasses = classNames(
    'terra-Form-control-label-text',
    labelTextAttrs.className,
  );

  const controlInputAttrs = Object.assign({}, inputAttrs);

  if (checked !== undefined) {
    controlInputAttrs.checked = checked;
  } else {
    controlInputAttrs.defaultChecked = defaultChecked;
  }

  return (
    <label htmlFor={id} {...customProps} className={labelClassNames}>
      <Input
        id={id}
        name={name}
        value={value}
        type={type}
        onChange={onChange}
        {...controlInputAttrs}
      />
      <span {...labelTextAttrs} className={labelTextClasses}>{labelText}</span>
    </label>
  );
};

Control.propTypes = propTypes;
Control.defaultProps = defaultProps;

export default Control;
