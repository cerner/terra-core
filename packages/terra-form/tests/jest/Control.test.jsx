/* eslint-disable react/jsx-boolean-value */

import React from 'react';
import Control from '../../src/Control';

it('should render a default checkbox when type is "checkbox"', () => {
  const choiceField = (<Control type="checkbox" />);
  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});

it('should render a default checkbox when type is "radio"', () => {
  const choiceField = (<Control type="radio" />);
  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});

it('should render a ChoiceField with a checkbox', () => {
  const choiceField = (
    <Control
      type="checkbox"
      labelText="Do you have any Children?"
      labelTextAttrs={{ className: 'healtheintent-label-text' }}
      name="children_present"
      value="children_present"
      required={true}
      inputAttrs={{ className: 'healtheintent-application' }}
      isInline={false}
    />);

  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});

it('should render a ChoiceField with a radio', () => {
  const choiceField = (
    <Control
      type="radio"
      labelText="Do you have any Children?"
      labelTextAttrs={{ className: 'healtheintent-label-text' }}
      name="children_present"
      value="children_present"
      required={true}
      inputAttrs={{ className: 'healtheintent-application' }}
      isInline={false}
    />);

  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});

it('should render as an uncontrolled radio when just a defaultChecked value is passed into the Control', () => {
  const choiceField = (<Control type="radio" defaultChecked />);
  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});

it('should render as an controlled radio when a checked value and onChange function are passed into the Control', () => {
  const choiceField = (<Control type="radio" checked onChange={() => {}} />);
  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});

it('should render as an uncontrolled checkbox when just a defaultChecked value is passed into the Control', () => {
  const choiceField = (<Control type="checkbox" defaultChecked />);
  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});

it('should render as an controlled checkbox when a checked value and onChange function are passed into the Control', () => {
  const choiceField = (<Control type="checkbox" checked onChange={() => {}} />);
  const wrapper = shallow(choiceField);
  expect(wrapper).toMatchSnapshot();
});
