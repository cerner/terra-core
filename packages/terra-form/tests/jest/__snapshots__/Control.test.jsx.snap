exports[`test should render a ChoiceField with a checkbox 1`] = `
<label
  className="terra-Form-control"
  required={true}>
  <Input
    className="healtheintent-application"
    name="children_present"
    required={false}
    type="checkbox"
    value="children_present" />
  <span
    className="terra-Form-control-label-text healtheintent-label-text">
    Do you have any Children?
  </span>
</label>
`;

exports[`test should render a ChoiceField with a radio 1`] = `
<label
  className="terra-Form-control"
  required={true}>
  <Input
    className="healtheintent-application"
    name="children_present"
    required={false}
    type="radio"
    value="children_present" />
  <span
    className="terra-Form-control-label-text healtheintent-label-text">
    Do you have any Children?
  </span>
</label>
`;

exports[`test should render a default checkbox when type is "checkbox" 1`] = `
<label
  className="terra-Form-control">
  <Input
    name={null}
    required={false}
    type="checkbox" />
  <span
    className="terra-Form-control-label-text" />
</label>
`;

exports[`test should render a default checkbox when type is "radio" 1`] = `
<label
  className="terra-Form-control">
  <Input
    name={null}
    required={false}
    type="radio" />
  <span
    className="terra-Form-control-label-text" />
</label>
`;

exports[`test should render as an controlled checkbox when a checked value and onChange function are passed into the Control 1`] = `
<label
  className="terra-Form-control">
  <Input
    checked={true}
    name={null}
    onChange={[Function]}
    required={false}
    type="checkbox" />
  <span
    className="terra-Form-control-label-text" />
</label>
`;

exports[`test should render as an controlled radio when a checked value and onChange function are passed into the Control 1`] = `
<label
  className="terra-Form-control">
  <Input
    checked={true}
    name={null}
    onChange={[Function]}
    required={false}
    type="radio" />
  <span
    className="terra-Form-control-label-text" />
</label>
`;

exports[`test should render as an uncontrolled checkbox when just a defaultChecked value is passed into the Control 1`] = `
<label
  className="terra-Form-control">
  <Input
    defaultChecked={true}
    name={null}
    required={false}
    type="checkbox" />
  <span
    className="terra-Form-control-label-text" />
</label>
`;

exports[`test should render as an uncontrolled radio when just a defaultChecked value is passed into the Control 1`] = `
<label
  className="terra-Form-control">
  <Input
    defaultChecked={true}
    name={null}
    required={false}
    type="radio" />
  <span
    className="terra-Form-control-label-text" />
</label>
`;
